### PR TITLE
VT Backed lectures and exercises

### PR DESCRIPTION
**Pull request by SlovenlyTF**
## Description

This pull request contains these completed PBI's:

[Web Create] As a content creator, I want to be able to put an exercise on the section I'm creating
[Web Create] As a content creator, I want to be able to put a lecture on the section I'm creating
[Web Edit] As a content creator, I want to be able to edit or delete sections, lectures or exercises of the courses I haven't published yet

## Changes

- Exercises
  - You can now create exercises.
  - You can now see a list of existing exercises.
  - You can now update details of an existing exercise.
  - You can now delete exercises.

- Lectures
  - You can now create lectures.
  - You can now see a list of existing lectures.
  - You can now update details of an existing lecture.
  - You can now delete lectures.

## Checklist

- [x] Code has been tested locally and passes all relevant tests.
- [x] Documentation has been updated to reflect the changes, if applicable.
- [x] Code follows the established coding style and guidelines of the project.
- [x] All new and existing tests related to the changes have passed.
- [x] Any necessary dependencies or new packages have been properly documented.
- [x] Pull request title and description are clear and descriptive.
- [x] Reviewers have been assigned to the pull request.
- [x] Any potential security implications have been considered and addressed.
- [x] Performance impact of the changes has been evaluated, if relevant.

### CODE REVIEW

File: routes/lectureRoutes.js
Comment (sofiiagran) [inline routes/lectureRoutes.js:1]: route: /getall/:id does not follow microservice guidelines.

 I would write: /section/:id so the full route would be api/lectures/section/:id

And please remove console.log

File: routes/exerciseRoutes.js
Comment (sofiiagran) [inline routes/exerciseRoutes.js:1]: please remove console.log as it is not necessary , and what is the / *   require login / * and why is it commented out?

**Review by sofiiagran — CHANGES_REQUESTED** [review]
Looks good, but please have a look at my (small) comments :)

File: routes/lectureRoutes.js
Comment (SlovenlyTF) [reply routes/lectureRoutes.js:1]: Yes, have fixed the route! :D

File: __tests__/routes/exercisesRoutes.test.js [hunk 1]
```diff
@@ -95,6 +95,50 @@ describe('PATCH: Update exercise route', () => {
...
+    exerciseCreate = await request(app)
+      .put('/api/exercises/' + fakeSection._id)
+      .set('Authorization', `Bearer ${token}`)
+      .send({ title: 'Test exercise', question: 'Test question', answers: {text: "test", correct: false, feedback: "test"} })
+      .expect(201);
+
+    expect(exerciseCreate.body.title).toBe('Test exercise');
+  });
+
+  console.log(exerciseCreate);
```
Comment (Vagn99) [inline __tests__/routes/exercisesRoutes.test.js:122]: Remove 🥰

File: routes/exerciseRoutes.js [hunk 2]
```diff
@@ -99,4 +100,39 @@ router.put("/:section_id", async (req, res) => {
...
+ * 
+ * @param {string} id - Exercise id
+ * @returns {string} - Just sends a message to confirm that the deletion is complete
+ */
+router.delete("/:id"/*, requireLogin*/, async (req, res) => {
+  const { id } = req.params; // destructure params
+
+  // Get the exercise object
+  const exercise = await ExerciseModel.findById(id).catch((err) => {
+    console.log(err)
```
Comment (Vagn99) [inline routes/exerciseRoutes.js:117]: Remove 🥰

File: routes/exerciseRoutes.js [hunk 3]
```diff
@@ -99,4 +100,39 @@ router.put("/:section_id", async (req, res) => {
...
+    res.status(204).send(err)
+  });
+
+  // Remove the exercise from the section exercises array
+  await SectionModel.updateOne({_id: exercise.parentSection}, {$pull: {exercises: exercise._id}}).catch((err) => {
+    console.log(err)
```
Comment (Vagn99) [inline routes/exerciseRoutes.js:123]: Remove 🥰

File: routes/exerciseRoutes.js [hunk 4]
```diff
@@ -99,4 +100,39 @@ router.put("/:section_id", async (req, res) => {
...
+    res.status(422).send({ error: errorCodes['E0012'] })
+  });
+
+  // Delete the exercise object
+  await ExerciseModel.findByIdAndDelete(id).catch((err) => {
+    console.log(err)
```
Comment (Vagn99) [inline routes/exerciseRoutes.js:129]: Remove 🥰

File: routes/lectureRoutes.js [hunk 5]
```diff
@@ -72,14 +71,28 @@ router.patch("/:id", /*requireLogin,*/ async (req, res) => {
 });
 
 
+/**
+ * Get all lectures from a specific section id
+ * @param {string} sid - section id
+ * @returns {object} - lectures
+ */
+router.get("/getall/:sid", async (req, res) => {
```
Comment (Vagn99) [inline routes/lectureRoutes.js:79]: Wrong route name! Should just be `GET /lectures/:id`

**Review by Vagn99 — CHANGES_REQUESTED** [review]
Just a few things, nothing big 🎉💯🙌🏽

File: routes/exerciseRoutes.js
Comment (SlovenlyTF) [reply routes/exerciseRoutes.js:1]: It's to check if the person is logged in when navigating the website.
The login token is not fetched everywhere, as the login was only just implemented by onboarding. We got an error when fetching the token, so we talked with onboarding, and they're setting up to fetch it everywhere.
When that is done, the requireLogin function will be uncommented👍

File: routes/lectureRoutes.js
Comment (SlovenlyTF) [reply routes/lectureRoutes.js:1]: The console.logs has now also been removed 👍

**Review by Vagn99 — APPROVED** [review]

File: routes/courseRoutes.js
Comment (sofiiagran) [inline routes/courseRoutes.js:1]: sorry to be annoying but where did you find the 422 status? it is not in the guidelines, should it be 404 maybe? and 204 on the delete course. Recommend to check the status code used by other groups and in the guidelines

**Review by sofiiagran — CHANGES_REQUESTED** [review]
see the comment about the use of 422, normally it should be 404 and 204 for delete requests... unless you have a good explanation?

File: routes/lectureRoutes.js (hunk 5 above)
Comment (magnemo) [reply routes/lectureRoutes.js]: Yes we have fixed it

File: routes/courseRoutes.js
Comment (sofiiagran) [reply routes/courseRoutes.js:1]: altså kig på det her delete request: 

router.delete("/:id", async (req, res) => {
  try {
    const { id } = req.params;

    const deletedCreator = await ContentCreator.findByIdAndDelete(id);

    if (!deletedCreator) {
      return res.status(204).json({ 'error': errorCodes['E0013'] });
    } else {
      return res.status(200).json({ message: "Content creator deleted successfully" });
    }
  } catch (error) {
    return res.status(500).json({ 'error': errorCodes['E0003'] });
  }
});

Comment (SlovenlyTF) [reply routes/courseRoutes.js:1]: Yes, you're right. It should be 204 with delete. It should be fixed now. Small oversight.

422 is unprocessable entity. It means that something in the input or code setup is wrong.

Comment (SlovenlyTF) [reply routes/courseRoutes.js:1]: Ah, du nåde lige at skrive inden jeg sendte.
204 synes jeg giver mere mening. Da vi prøver at finde en ting som vi vil slette, hvis vi ikke kan finde den, betyder det at den ikke eksistere, hvilket er godt, men siden det ikke var måden vi forventede, skal den ikke give direkte 200 (ok)

File: routes/courseRoutes.js [hunk 6]
```diff
@@ -307,21 +305,17 @@ router.delete("/:id"/*, requireLogin*/, async (req, res) => {
     lectureIds.map(async (lecture_id) => {
 
       // Delete the lecture
-      await LectureModel.findByIdAndDelete( lecture_id, (err) => {
+      await LectureModel.findByIdAndDelete(lecture_id).catch((err) => res.status(422).send(err));
```
Comment (MortenMunk) [inline routes/courseRoutes.js:308]: Is 422 the correct response status?

The API guidelines specify this:
_"DO return a 204-No Content without a resource/body for a DELETE operation (even if the URL identifies a resource that does not exist; do not return 404-Not Found)"_

I think you should double check this, as I am not too sure.

File: routes/courseRoutes.js [hunk 7]
```diff
@@ -307,21 +305,17 @@ router.delete("/:id"/*, requireLogin*/, async (req, res) => {
...
 
-      });
     });
 
     // Delete the section
-    await SectionModel.findByIdAndDelete( section_id , (err) => {
-
-
-    });
+    await SectionModel.findByIdAndDelete(section_id).catch((err) => res.status(422).send(err));
```
Comment (MortenMunk) [inline routes/courseRoutes.js:313]: Same as above, should this be 422?

File: routes/courseRoutes.js [hunk 8]
```diff
@@ -307,21 +305,17 @@ router.delete("/:id"/*, requireLogin*/, async (req, res) => {
...
   });
 
   // Delete the course
-  await CourseModel.findByIdAndDelete( id , (err) => {
-  });
+  await CourseModel.findByIdAndDelete(id).catch((err) => res.status(422).send(err));
```
Comment (MortenMunk) [inline routes/courseRoutes.js:317]: Same as above, should this be 422?

File: routes/exerciseRoutes.js [hunk 9]
```diff
@@ -77,10 +77,7 @@ router.put("/:section_id", async (req, res) => {
       },
       function (err, docs) {
         if (err) {
-
-          res.send(err);
-        } else {
-
+          res.status(422).send(err);
```
Comment (MortenMunk) [inline routes/exerciseRoutes.js:80]: error status 422 may make sense for a PATCH operation, but I am not sure, so maybe this should be double checked.

File: routes/lectureRoutes.js [hunk 10]
```diff
@@ -86,7 +84,7 @@ router.patch("/:id", /*requireLogin,*/ async (req, res) => {
     },
     function (err, docs) {
       if (err) {
-        res.send(err);
+        res.status(422).send(err);
```
Comment (MortenMunk) [inline routes/lectureRoutes.js:87]: Same as above

File: routes/lectureRoutes.js [hunk 11]
```diff
@@ -119,15 +117,12 @@ router.delete("/:id"/*, requireLogin*/, async (req, res) => {
...
-
-  });
+  const lecture = await LectureModel.findById(id).catch((err) => res.status(204).send(err));
 
   // Get the section object
   const section_id = lecture.parentSection;
-  const section = await SectionModel.findById(section_id).catch((err) => {
-
-  });
+  const section = await SectionModel.findById(section_id).catch((err) => res.status(422).send(err));
```
Comment (MortenMunk) [inline routes/lectureRoutes.js:124]: is 422 correct status?

File: routes/lectureRoutes.js [hunk 12]
```diff
@@ -144,7 +139,7 @@ router.delete("/:id"/*, requireLogin*/, async (req, res) => {
 
   // Delete the lecture object
   await LectureModel.findByIdAndDelete(id).catch((err) => {
-
+    res.status(422).send(err);
```
Comment (MortenMunk) [inline routes/lectureRoutes.js:142]: same as above

**Review by MortenMunk — CHANGES_REQUESTED** [review]

**Review by sofiiagran — APPROVED** [review]
As morten I am unsure about 422, but I will approve for now and then when making PR to staging we should do some research

**Review by MortenMunk — APPROVED** [review]
Looks good to me.. Maybe 422 status codes should be replaced, but this is up for debate.
Approving 👍 🥇

**Comment by SlovenlyTF** [comment]
The last 422 has been replaced with 400

**Pull request merged**